src/runtime/hexagon/hexagon_module.cc

```diff
@@ -0,0 +1,296 @@
+/*!
+ * Copyright 2018-2019 Qualcomm Innovation Center, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE‐2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * This Contribution is being provided by Qualcomm Technologies, Inc.,
+ * a Delaware corporation, or its subsidiary Qualcomm Innovation Center, Inc.,
+ * a California corporation, under certain additional terms and conditions
+ * pursuant to Section 5 of the Apache 2.0 license.  In this regard, with
+ * respect to this Contribution, the term "Work" in Section 1 of the
+ * Apache 2.0 license means only the specific subdirectory within the TVM repo
+ * (currently at https://github.com/dmlc/tvm) to which this Contribution is
+ * made.
+ * In any case, this submission is "Not a Contribution" with respect to its
+ * permitted use with any of the "vta" and "verilog" subdirectories in the TVM
+ * repo.
+ */
+#include "hexagon_module.h"
+
+#include <tvm/runtime/registry.h>
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "../file_util.h"
+#include "../meta_data.h"
+#include "hexagon_device_sim.h"
+#include "hexagon_device_target.h"
+#ifdef __ANDROID__
+#include <android/log.h>
+#endif
+
+namespace tvm {
+namespace runtime {
+
+hexagon::Device::~Device() {}
+
+namespace hexagon {
+
+Hardware HW;
```

Comment:
When is this constructed or destructed? Does it work when accessing in multiple threads?

Reply:
It's constructed when the library libtvm_runtime.so is loaded and it's destroyed when it's unloaded.  We need a persistent object representing the hardware, do you have any concerns about having it as a global variable?

We haven't considered thread-safety. We'll do it, but we need more information about the use of threads in the TVM runtime.

Reply:
Replaced with a singleton pattern.